services/leader/client.go

```diff
@@ -0,0 +1,239 @@
+	// with no leader. We duplicate this error so the user doesn't have to
+	// import etcd's concurrency package in order to check the cause of the
+	// error.
+	ErrNoLeader = concurrency.ErrElectionNoLeader
+
+	// ErrSessionExpired is returned when a client's session (etcd lease) is no
+	// longer being refreshed for any reason (due to expiration, error state,
+	// etc.).
+	ErrSessionExpired = errors.New("election client session (lease) expired")
+)
+
+// NB(mschalle): when an etcd leader failover occurs, all current leases have
+// their TTLs refreshed: https://github.com/coreos/etcd/issues/2660
+
+type client struct {
+	sync.Mutex
+
+	ctxCancel   context.CancelFunc
+	election    *concurrency.Election
+	session     *concurrency.Session
+	opts        services.ElectionOptions
+	closed      uint32
+	campaigning uint32
+	wb          xwatch.Watchable
+	w           xwatch.Watch
+}
+
+// newClient returns an instance of an client client bound to a single election.
+func newClient(cli *clientv3.Client, opts Options, electionID string, ttl int) (*client, error) {
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	var sessionOpts []concurrency.SessionOption
+	if ttl != 0 {
+		sessionOpts = append(sessionOpts, concurrency.WithTTL(ttl))
+	}
+
+	session, err := concurrency.NewSession(cli, sessionOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	election := concurrency.NewElection(session, electionPrefix(opts.ServiceID(), electionID))
+
+	wb := xwatch.NewWatchable()
+	wb.Update(okCampaignStatus(CampaignFollower))
+
+	_, w, err := wb.Watch()
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		election: election,
+		session:  session,
+		opts:     opts.ElectionOpts(),
+		wb:       wb,
+		w:        w,
+	}, nil
+}
+
+// val returns the value the leader should propose based on (1) a potentially
+// empty override value and (2) the hostname of the caller (with a fallback to
+// the DefaultHostname option).
+func (c *client) val(override string) string {
+	if override != "" {
+		return override
+	}
+
+	if h, err := os.Hostname(); err == nil {
+		return h
+	}
+
+	return c.opts.DefaultHostname()
+}
+
+func (c *client) campaign(override string) (xwatch.Watch, error) {
+	if c.isClosed() {
+		return nil, errClientClosed
+	}
+
+	if !atomic.CompareAndSwapUint32(&c.campaigning, 0, 1) {
+		return nil, ErrCampaignInProgress
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	c.Lock()
+	c.ctxCancel = cancel
+	c.Unlock()
+
+	go func() {
+		c.wb.Update(okCampaignStatus(CampaignFollower))
+		// blocks until elected or error
+		err := c.election.Campaign(ctx, c.val(override))
+		if err != nil {
+			c.wb.Update(errCampaignStatus(err))
+		}
+
+		select {
+		case <-c.session.Done():
+			c.wb.Update(errCampaignStatus(ErrSessionExpired))
+		default:
+			c.wb.Update(okCampaignStatus(CampaignLeader))
+		}
+
+		cancel()
+		c.Lock()
+		c.ctxCancel = nil
+		c.Unlock()
+	}()
+
+	return c.w, nil
```

Comment:
should create a new watch from the watchable and return it, otherwise if the watch is shared by multiple users and one of them closed it, other users will be affected